influence_benchmark/RL/expert_iteration.py

```diff
@@ -0,0 +1,217 @@
+                        "preferences": state.preferences,
+                        "transition_probs": state.transition_probs,
+                    }
+                )
+
+        env_trajectories = [
+            [traj for traj in trajectories if int(traj["trajectory_id"]) < start_trajectory_id + num_trajectories]
+            for trajectories in env_trajectories
+        ]
+
+        save_path = trajectory_folder / f"{device.split(':')[-1]}.jsonl"
+        save_path.parent.mkdir(parents=True, exist_ok=True)
+        with open(save_path, "w", encoding="utf-8") as f:
+            for env in env_trajectories:
+                for turn_data in env:
+                    f.write(json.dumps(turn_data) + "\n")
+
+        backend.close()
+
+    def rank_trajectories_by_avg_reward(self, trajectory_folder):
+        trajectories = []
+        for file in trajectory_folder.iterdir():
+            if file.name[0] in [str(x) for x in range(10)]:
+                with open(file, "r", encoding="utf-8") as f:
+                    trajectories_in_file = [json.loads(line) for line in f]
+                    trajectories.extend(trajectories_in_file)
+        trajectory_rewards = defaultdict(list)
+        for trajectory in trajectories:
+            trajectory_id = trajectory["trajectory_id"]
+            expected_preference = 0
+            for key, value in trajectory["preferences"].items():
+                expected_preference += int(key) * value
+            trajectory_rewards[trajectory_id].append(expected_preference)
+
+        avg_rewards = {key: sum(value) / len(value) for key, value in trajectory_rewards.items()}
+        sorted_trajectories = sorted(trajectories, key=lambda x: avg_rewards[x["trajectory_id"]], reverse=True)
+        num_selected = 0
+        selected_trajectories = []
+        selected_trajectory_ids = set()
+        for trajectory in sorted_trajectories:
+            selected_trajectories.append(trajectory)
+            if trajectory["trajectory_id"] not in selected_trajectory_ids:
+                num_selected += 1
+                if num_selected > self.num_chosen_trajectories:
+                    break
+                selected_trajectory_ids.add(trajectory["trajectory_id"])
+
+        return selected_trajectories
+
+    def format_and_save_trajectories_for_SFT(self, selected_trajectories, trajectory_folder):
+        formatted_trajectories = []
+        for trajectory in selected_trajectories:
+            system_prompt = trajectory["agent_system_prompt"][0]["content"]
+            messages = [{"role": "system", "content": system_prompt}]
+            messages.extend(
+                [
+                    (
+                        {"role": "assistant", "content": msg["content"]}
+                        if msg["role"] == "agent"
+                        else {"role": "user", "content": msg["content"]}
+                    )
+                    for msg in trajectory["history"]
```

Comment:
This is going to be updated to not train on the initial messages, right?

Reply:
yes this is fixed now